app/controllers/application_controller.rb

```diff
@@ -22,6 +22,39 @@ class ApplicationController < ActionController::API
 
   private
 
+  set_current_tenant_through_filter
+  before_action :set_the_current_tenant
+  after_action  :unset_the_current_tenant
+
+  def set_the_current_tenant
+    return unless Tenant.tenancy_enabled?
+
+    tenant = Tenant.find_by(:external_tenant => user_identity)
+    if tenant
+      set_current_tenant(tenant)
+    else
+      error_document = {
+          "errors" => [
+              {
+                  "status" => "401",
+                  "detail" => "Account number not present or not found"
+              }
+          ]
+      }
+      render :json => error_document, :status => :unauthorized
+    end
+  end
+
+  def unset_the_current_tenant
+    set_current_tenant(nil)
+  end
+
+  def user_identity
+    ManageIQ::API::Common::Request.with_request(request) do |current|
+      current.user.tenant if current.headers.key?(ManageIQ::API::Common::Request::IDENTITY_KEY)
```

Comment:
checking the key this way may not be case-insensitive, so this might not be a accurate.  Perhaps just get the current.user.tenant and catch the exception instead?  It should be set, so exception handling feels like the right path.  Additionally, if there is no tenant, what happens?  This code doesn't raise, so it just moves on?